pumps/resurface.go

```diff
@@ -0,0 +1,217 @@
+	return resurfacePumpName
+}
+
+func (rp *ResurfacePump) GetEnvPrefix() string {
+	return rp.config.EnvPrefix
+}
+
+func (rp *ResurfacePump) Init(config interface{}) error {
+	rp.config = &ResurfacePumpConfig{}
+	rp.log = log.WithField("prefix", resurfacePrefix)
+
+	err := mapstructure.Decode(config, &rp.config)
+	if err != nil {
+		rp.log.Fatal("Failed to decode configuration: ", err)
+	}
+
+	processPumpEnvVars(rp, rp.log, rp.config, resurfaceDefaultEnv)
+
+	opt := logger.Options{
+		Rules: rp.config.Rules,
+		Url:   rp.config.Url,
+		Queue: rp.config.Queue,
+	}
+	logger, err := logger.NewHttpLogger(opt)
+	if err != nil {
+		log.Fatal(err)
+	}
+	rp.logger = logger
+	if logger.Enabled() {
+		rp.log.Info(rp.GetName() + " Initialized")
+		return nil
+	} else {
+		rp.log.Info(rp.GetName() + " Initialized (Logger disabled)")
+		return errors.New("Logger is not enabled")
+	}
+}
+
+func parseHeaders(headersString string, existingHeaders http.Header) http.Header {
+	var headers http.Header
+	if existingHeaders != nil {
+		headers = http.Header.Clone(existingHeaders)
+	} else {
+		headers = http.Header{}
+	}
+	for _, line := range strings.Split(headersString, "\r\n") {
+		header := strings.Split(line, ": ")
+		if len(header) < 2 {
+			continue
+		}
+		headers.Add(header[0], header[1])
+	}
+	return headers
+}
+
+func mapRawData(rec analytics.AnalyticsRecord) (http.Request, http.Response, map[string]string) {
+	var req [3]string
+	var res [3]string
+	tykFields := [6]string{
+		"API-ID",
+		"API-Key",
+		"API-Name",
+		"API-Version",
+		"Oauth-ID",
+		"Org-ID",
+	}
+
+	// Decode raw HTTP transaction from base64 strings
+	rawBytesReq, err := base64.StdEncoding.DecodeString(rec.RawRequest)
+	if err != nil {
+		log.Fatal(err)
+	}
+	rawBytesRes, err := base64.StdEncoding.DecodeString(rec.RawResponse)
+	if err != nil {
+		log.Fatal(err)
+	}
+	rawReq := string(rawBytesReq)
+	rawRes := string(rawBytesRes)
+
+	// Slice first line, headers, body+trailers
+	copy(req[:2], strings.SplitN(rawReq, "\r\n", 2))
+	copy(res[:2], strings.SplitN(rawRes, "\r\n", 2))
+	copy(req[1:], strings.SplitN(req[1], "\r\n\r\n", 2))
+	copy(res[1:], strings.SplitN(res[1], "\r\n\r\n", 2))
+
+	// Request method
+	method := strings.Fields(req[0])[0]
+	if method == "" {
+		method = rec.Method
+	}
+
+	// Request headers
+	reqHeaders := parseHeaders(req[1], nil)
+
+	// Request address
+	if reqHeaders.Get("X-FORWARDED-FOR") == "" {
+		reqHeaders.Add("X-FORWARDED-FOR", rec.IPAddress)
+	}
+
+	// Request URL
+	schema := "http"               // TODO - how to determine schema from Tyk Analytics?
+	host := reqHeaders.Get("Host") // TODO - how to determine actual used hostname? (not Host header)
+	if host == "" {
+		host = rec.Host
+	}
+	path := strings.Fields(req[0])[1]
+	if path == "" {
+		path = rec.Path
+	}
+	if !strings.HasPrefix(path, "/") {
+		path = "/" + path
+	}
+	parsedURL, err := url.Parse(schema + "://" + host + path)
+	if err != nil {
+		log.Fatal("Problem converting url", err)
```

Comment:
@monrax can we remove the `log.Fatal` from here too? Why do we need to parse the URL?

Reply:
Thank you for pointing it out, I just refactored all error handling.

Internally, the Resurface backend uses our [logger-go](https://github.com/resurfaceio/logger-go) tool which takes care of parsing all available details from the HTTP transaction. In addition, it applies some [logging rules](https://resurface.io/docs#logging-rules), and submits the entire reconstructed message to the configured [Capture URL ](https://resurface.io/docs#getting-capture-url) to the running Resurface instance using a specific [JSON format](https://resurface.io/docs#json-format). The main use cases of the `logger-go` package are middleware implementations, so we chose both the `Request` and `Response` types from the `net/http` package as the vehicles to receive HTTP transaction data but, as long as the data can be added to those structs, `logger-go` can be used in other types of implementations (like a Tyk Pump backend!) just fine. All this is to say, the `URL` field in the `Request` struct is a [URL](https://pkg.go.dev/net/url#URL) type from the `net/url` package, so it has to be parsed in order to assign it to `httpReq.URL`.